### ISSUE TITLE
blead hangs on FreeBSD during testing

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11442]
Migrated from [rt.perl.org#92828](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=92828) (status was 'resolved')

Searchable as RT92828$

**Comment by p5pRT** [comment]
### From 0briang0@gmail.com

#### Created by briang@cpan.org



My FreeBSD blead smoker has been hanging at test 6 in t/io/eintr\.t for
the past few days\. git bisect tracked it down to commit
df375c6d048b938ecdeaecc7b264a7f1a190120a \(\[perl \#8611\] eintr\.t&#8203;: skip
based on capability rather than OS\)

<details><summary>Perl Info</summary>

```

Flags:
   category=install
   severity=low

Site configuration information for perl 5.15.0:

Configured by bri at Tue Jun 14 14:35:20 BST 2011.

Summary of my perl5 (revision 5 version 15 subversion 0) configuration:
 Commit id: e528b6072f9f5a668f9b01d9ab00c92a31d12e64
 Platform:
   osname=freebsd, osvers=8.1-release, archname=amd64-freebsd
   uname='freebsd bsd.zombie.org.uk 8.1-release freebsd 8.1-release
#0: mon jul 19 02:36:49 utc 2010
root@mason.cse.buffalo.edu:usrobjusrsrcsysgeneric amd64 '
   config_args='-des -Dusedevel'
   hint=recommended, useposix=true, d_sigaction=define
   useithreads=undef, usemultiplicity=undef
   useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
   use64bitint=define, use64bitall=define, uselongdouble=undef
   usemymalloc=n, bincompat5005=undef
 Compiler:
   cc='cc', ccflags ='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include',
   optimize='-O',
   cppflags='-DHAS_FPSETMASK -DHAS_FLOATINGPOINT_H
-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
   ccversion='', gccversion='4.2.1 20070719  [FreeBSD]', gccosandvers=''
   intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
   d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
   ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t',
lseeksize=8
   alignbytes=8, prototype=define
 Linker and Libraries:
   ld='cc', ldflags ='-Wl,-E  -fstack-protector -L/usr/local/lib'
   libpth=/usr/lib /usr/local/lib
   libs=-lgdbm -lm -lcrypt -lutil -lc
   perllibs=-lm -lcrypt -lutil -lc
   libc=, so=so, useshrplib=false, libperl=libperl.a
   gnulibc_version=''
 Dynamic Linking:
   dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
   cccdlflags='-DPIC -fPIC', lddlflags='-shared  -L/usr/local/lib
-fstack-protector'

Locally applied patches:



@INC for perl 5.15.0:
   lib
   /home/bri/lib/perl
   /usr/local/lib/perl5/site_perl/5.15.0/amd64-freebsd
   /usr/local/lib/perl5/site_perl/5.15.0
   /usr/local/lib/perl5/5.15.0/amd64-freebsd
   /usr/local/lib/perl5/5.15.0
   /usr/local/lib/perl5/site_perl
   .


Environment for perl 5.15.0:
   HOME=/home/bri
   LANG=C
   LANGUAGE (unset)
   LD_LIBRARY_PATH (unset)
   LOGDIR (unset)
   PATH=/home/bri/bin:/sbin:/bin:/usr/sbin:/usr/bin:/usr/games:/usr/local/sbin:/usr/local/bin:/home/bri/bin
   PERL5LIB=/home/bri/lib/perl
   PERL_BADLANG (unset)
   SHELL=/usr/local/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @xdg

That commit changed from a hard\-coded list of OSes to skip to a probe to
see if things are interruptable\.

Unless there is good reason for the change\, I think we should just revert\.

Dave M \-\- as it was your commit\, I'm assigning this ticket to you to
resolve the FreeBSD issue or please revert the commit\.

Thanks\,

David G \(RMOTM\)

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Jun 14\, 2011 at 08&#8203;:43&#8203;:08AM \-0700\, brian greenfield wrote&#8203;:

> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
> 
> My FreeBSD blead smoker has been hanging at test 6 in t/io/eintr\.t for
> the past few days\. git bisect tracked it down to commit
> df375c6d048b938ecdeaecc7b264a7f1a190120a \(\[perl \#8611\] eintr\.t&#8203;: skip
> based on capability rather than OS\)

I see this\. The relevant bits of kdump seem to be&#8203;:

34054 perl     GIO   fd 1 wrote 30 bytes
&nbsp;       "ok 5 \- read/die&#8203;: close status
&nbsp;       "
 34054 perl     RET   write 30/0x1e
 34054 perl     CALL  close\(0x4\)
 34054 perl     RET   close 0
 34054 perl     CALL  pipe
 34054 perl     RET   pipe 3
 34054 perl     CALL  ioctl\(0x3\,TIOCGETA\,0x7fffffffe750\)
 34054 perl     RET   ioctl \-1 errno 25 Inappropriate ioctl for device
 34054 perl     CALL  lseek\(0x3\,0\,SEEK\_CUR\)
 34054 perl     RET   lseek \-1 errno 29 Illegal seek
 34054 perl     CALL  ioctl\(0x4\,TIOCGETA\,0x7fffffffe750\)
 34054 perl     RET   ioctl \-1 errno 25 Inappropriate ioctl for device
 34054 perl     CALL  lseek\(0x4\,0\,SEEK\_CUR\)
 34054 perl     RET   lseek \-1 errno 29 Illegal seek
 34054 perl     CALL  fcntl\(0x3\,F\_SETFD\,FD\_CLOEXEC\)
 34054 perl     RET   fcntl 0
 34054 perl     CALL  fcntl\(0x4\,F\_SETFD\,FD\_CLOEXEC\)
 34054 perl     RET   fcntl 0
 34054 perl     CALL  sigprocmask\(SIG\_BLOCK\,0x7fffffffe8a0\,0x7fffffffe910\)
 34054 perl     RET   sigprocmask 0
 34054 perl     CALL  sigaction\(SIGALRM\,0x7fffffffe890\,0x7fffffffe870\)
 34054 perl     RET   sigaction 0
 34054 perl     CALL  sigprocmask\(SIG\_SETMASK\,0x7fffffffe730\,0\)
 34054 perl     RET   sigprocmask 0
 34054 perl     CALL  mmap\(0\,0x100000\,PROT\_READ|PROT\_WRITE\,MAP\_PRIVATE|MAP\_ANON\,
0xffffffff\,0\)
 34054 perl     RET   mmap 16777216/0x801000000
 34054 perl     CALL  mmap\(0\,0x100000\,PROT\_READ|PROT\_WRITE\,MAP\_PRIVATE|MAP\_ANON\,
0xffffffff\,0\)
 34054 perl     RET   mmap 17825792/0x801100000
 34054 perl     CALL  setitimer\(0\,0x7fffffffe9d0\,0x7fffffffe9b0\)
 34054 perl     RET   setitimer 0
 34054 perl     CALL  write\(0x4\,0x800f40000\,0x2000\)
 34054 perl     GIO   fd 4 wrote 4096 bytes
&nbsp;       "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa\\

\.\.\.

&nbsp;        aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
 34054 perl     RET   write 8192/0x2000
 34054 perl     PSIG  SIGALRM caught handler=0x4b8740 mask=0x0 code=0x0
 34054 perl     CALL  sigreturn\(0x7fffffffe260\)
 34054 perl     RET   sigreturn JUSTRETURN
 34054 perl     CALL  write\(0x4\,0x800f40000\,0x2000\)


ie a write to a pipe is restarted\.

Except\, how come it has reached test 5 before a write restarts?
Shouldn't the earlier probe bail out?

Nicholas Clark

**Comment by p5pRT** [comment]
### From @craigberry

On Tue\, Jun 14\, 2011 at 4&#8203;:10 PM\, Nicholas Clark \<nick@&#8203;ccl4\.org> wrote&#8203;:
> On Tue\, Jun 14\, 2011 at 08&#8203;:43&#8203;:08AM \-0700\, brian greenfield wrote&#8203;:
>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> \[Please describe your issue here\]
>>
>> My FreeBSD blead smoker has been hanging at test 6 in t/io/eintr\.t for
>> the past few days\. git bisect tracked it down to commit
>> df375c6d048b938ecdeaecc7b264a7f1a190120a \(\[perl \#8611\] eintr\.t&#8203;: skip
>> based on capability rather than OS\)


> ie a write to a pipe is restarted\.
>
> Except\, how come it has reached test 5 before a write restarts?
> Shouldn't the earlier probe bail out?

The comments in the test script indicate&#8203;:

 \# On FreeBSD\, writes to pipes of 8192 bytes or more use a mechanism
 \# that is not interruptible \(see perl \#85842 and \#84688\)\.

I think the probe just opens the pipe and doesn't actually move data
through it\, especially not a chunk of data large enough to trigger the
FreeBSD\-specific pipe problem\.

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Jun 14\, 2011 at 10&#8203;:10&#8203;:36PM \-0500\, Craig A\. Berry wrote&#8203;:
> On Tue\, Jun 14\, 2011 at 4&#8203;:10 PM\, Nicholas Clark \<nick@&#8203;ccl4\.org> wrote&#8203;:
> > On Tue\, Jun 14\, 2011 at 08&#8203;:43&#8203;:08AM \-0700\, brian greenfield wrote&#8203;:
> >
> >> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> >> \[Please describe your issue here\]
> >>
> >> My FreeBSD blead smoker has been hanging at test 6 in t/io/eintr\.t for
> >> the past few days\. git bisect tracked it down to commit
> >> df375c6d048b938ecdeaecc7b264a7f1a190120a \(\[perl \#8611\] eintr\.t&#8203;: skip
> >> based on capability rather than OS\)
> 
> 
> > ie a write to a pipe is restarted\.
> >
> > Except\, how come it has reached test 5 before a write restarts?
> > Shouldn't the earlier probe bail out?
> 
> The comments in the test script indicate&#8203;:
> 
>  \# On FreeBSD\, writes to pipes of 8192 bytes or more use a mechanism
>  \# that is not interruptible \(see perl \#85842 and \#84688\)\.
> 
> I think the probe just opens the pipe and doesn't actually move data
> through it\, especially not a chunk of data large enough to trigger the
> FreeBSD\-specific pipe problem\.

Now fixed \(hopefully\) by the following&#8203;:

commit 6e59d93aee950461947904b4f24a7f52c6c85f58
Author&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
AuthorDate&#8203;: Wed Jun 15 15&#8203;:07&#8203;:57 2011 \+0100
Commit&#8203;:     David Mitchell \<davem@&#8203;iabyn\.com>
CommitDate&#8203;: Wed Jun 15 15&#8203;:11&#8203;:27 2011 \+0100

&nbsp;    \[perl \#92828\] eintr\.t hangs on FreeBSD
&nbsp;    
&nbsp;    My commit df375c6d048b938ecdeaecc7b264a7f1a190120a attempted to
&nbsp;    convert t/io/eintr\.t from OS\-based skipping to capability\-based skipping\,
&nbsp;    but it only tested whether reads from pipes are interruptible\.
&nbsp;    Some OSes \(like FreeBSD\) only hang on write; so probe for that too\.


Affected files \.\.\.
&nbsp;    
&nbsp;    M    t/io/eintr\.t


\-\- 
"You may not work around any technical limitations in the software"
&nbsp;    \-\- Windows Vista license

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From @tomhukins

On Wed\, Jun 15\, 2011 at 03&#8203;:12&#8203;:29PM \+0100\, Dave Mitchell wrote&#8203;:
> On Tue\, Jun 14\, 2011 at 10&#8203;:10&#8203;:36PM \-0500\, Craig A\. Berry wrote&#8203;:
> > On Tue\, Jun 14\, 2011 at 4&#8203;:10 PM\, Nicholas Clark \<nick@&#8203;ccl4\.org> wrote&#8203;:
> > > On Tue\, Jun 14\, 2011 at 08&#8203;:43&#8203;:08AM \-0700\, brian greenfield wrote&#8203;:
> > >
> > >> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > >> \[Please describe your issue here\]
> > >>
> > >> My FreeBSD blead smoker has been hanging at test 6 in t/io/eintr\.t for
> > >> the past few days\. git bisect tracked it down to commit
> > >> df375c6d048b938ecdeaecc7b264a7f1a190120a \(\[perl \#8611\] eintr\.t&#8203;: skip
> > >> based on capability rather than OS\)
> > 
> > 
> > > ie a write to a pipe is restarted\.
> > >
> > > Except\, how come it has reached test 5 before a write restarts?
> > > Shouldn't the earlier probe bail out?
> > 
> > The comments in the test script indicate&#8203;:
> > 
> >  \# On FreeBSD\, writes to pipes of 8192 bytes or more use a mechanism
> >  \# that is not interruptible \(see perl \#85842 and \#84688\)\.
> > 
> > I think the probe just opens the pipe and doesn't actually move data
> > through it\, especially not a chunk of data large enough to trigger the
> > FreeBSD\-specific pipe problem\.
> 
> Now fixed \(hopefully\) by the following&#8203;:
> 
> commit 6e59d93aee950461947904b4f24a7f52c6c85f58

This test still hangs for me in the same way as before\.  The kdump
output looks much like the one Nicholas posted earlier\.

Tom

**Comment by p5pRT** [comment]
### From @nwc10

On Wed\, Jun 15\, 2011 at 03&#8203;:59&#8203;:25PM \+0100\, Tom Hukins wrote&#8203;:
> On Wed\, Jun 15\, 2011 at 03&#8203;:12&#8203;:29PM \+0100\, Dave Mitchell wrote&#8203;:
> > On Tue\, Jun 14\, 2011 at 10&#8203;:10&#8203;:36PM \-0500\, Craig A\. Berry wrote&#8203;:
> > > On Tue\, Jun 14\, 2011 at 4&#8203;:10 PM\, Nicholas Clark \<nick@&#8203;ccl4\.org> wrote&#8203;:
> > > > On Tue\, Jun 14\, 2011 at 08&#8203;:43&#8203;:08AM \-0700\, brian greenfield wrote&#8203;:
> > > >
> > > >> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > > >> \[Please describe your issue here\]
> > > >>
> > > >> My FreeBSD blead smoker has been hanging at test 6 in t/io/eintr\.t for
> > > >> the past few days\. git bisect tracked it down to commit
> > > >> df375c6d048b938ecdeaecc7b264a7f1a190120a \(\[perl \#8611\] eintr\.t&#8203;: skip
> > > >> based on capability rather than OS\)
> > > 
> > > 
> > > > ie a write to a pipe is restarted\.
> > > >
> > > > Except\, how come it has reached test 5 before a write restarts?
> > > > Shouldn't the earlier probe bail out?
> > > 
> > > The comments in the test script indicate&#8203;:
> > > 
> > >  \# On FreeBSD\, writes to pipes of 8192 bytes or more use a mechanism
> > >  \# that is not interruptible \(see perl \#85842 and \#84688\)\.
> > > 
> > > I think the probe just opens the pipe and doesn't actually move data
> > > through it\, especially not a chunk of data large enough to trigger the
> > > FreeBSD\-specific pipe problem\.
> > 
> > Now fixed \(hopefully\) by the following&#8203;:
> > 
> > commit 6e59d93aee950461947904b4f24a7f52c6c85f58
> 
> This test still hangs for me in the same way as before\.  The kdump
> output looks much like the one Nicholas posted earlier\.

Sadly\, me too\, on FreeBSD 7\.2&#8203;:

Summary of my perl5 \(revision 5 version 15 subversion 0\) configuration&#8203;:
&nbsp;  Commit id&#8203;: 6e59d93aee950461947904b4f24a7f52c6c85f58
&nbsp;  Platform&#8203;:
&nbsp;    osname=freebsd\, osvers=7\.2\-release\, archname=amd64\-freebsd\-thread\-multi
&nbsp;    uname='freebsd spectre\.mongueurs\.net 7\.2\-release freebsd 7\.2\-release \#0&#8203;: fri may 1 07&#8203;:18&#8203;:07 utc 2009 root@&#8203;driscoll\.cse\.buffalo\.edu&#8203;:usrobjusrsrcsysgeneric amd64 '
&nbsp;    config\_args='\-Dgccansipedantic \-Accflags=\-ansi \-pedantic \-Dusedevel=y \-Dcc=ccache gcc \-Dld=gcc \-Ubincompat5005 \-Uinstallusrbinperl \-Dcf\_email=nick@&#8203;ccl4\.org \-Dperladmin=nick@&#8203;ccl4\.org \-Dinc\_version\_list=  \-Dinc\_version\_list\_init=0 \-Doptimize=\-g \-Dusethreads \-Duse64bitint \-Accflags=\-DNO\_MATHOMS \-Uusemymalloc \-Duseperlio \-Dprefix=~/Sandpit/snap5\.9\.x\- \-Dinstallman1dir=none \-Dinstallman3dir=none \-Uuserelocatableinc \-de'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=define\, use64bitall=define\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='ccache gcc'\, ccflags ='\-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-ansi \-pedantic \-DNO\_MATHOMS \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'\,
&nbsp;    optimize='\-g'\,
&nbsp;    cppflags='\-DHAS\_FPSETMASK \-DHAS\_FLOATINGPOINT\_H \-ansi \-pedantic \-DNO\_MATHOMS \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.2\.1 20070719  \[FreeBSD\]'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=8\, ptrsize=8\, doublesize=8\, byteorder=12345678
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=8\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags ='\-pthread \-Wl\,\-E  \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/lib /usr/local/lib
&nbsp;    libs=\-lgdbm \-lm \-lcrypt
&nbsp;    perllibs=\-lm \-lcrypt
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags='\-DPIC \-fPIC'\, lddlflags='\-shared  \-L/usr/local/lib \-fstack\-protector'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: MULTIPLICITY NO\_MATHOMS PERL\_DONT\_CREATE\_GVSV
&nbsp;                        PERL\_IMPLICIT\_CONTEXT PERL\_MALLOC\_WRAP
&nbsp;                        PERL\_PRESERVE\_IVUV PERL\_USE\_DEVEL USE\_64\_BIT\_ALL
&nbsp;                        USE\_64\_BIT\_INT USE\_ITHREADS USE\_LARGE\_FILES
&nbsp;                        USE\_LOCALE USE\_LOCALE\_COLLATE USE\_LOCALE\_CTYPE
&nbsp;                        USE\_LOCALE\_NUMERIC USE\_PERLIO USE\_PERL\_ATOF
&nbsp;                        USE\_REENTRANT\_API
&nbsp;  Built under freebsd
&nbsp;  Compiled at Jun 15 2011 16&#8203;:26&#8203;:46
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    lib
&nbsp;    /home/nicholas/Sandpit/snap5\.9\.x\-/lib/perl5/site\_perl/5\.15\.0/amd64\-freebsd\-thread\-multi
&nbsp;    /home/nicholas/Sandpit/snap5\.9\.x\-/lib/perl5/site\_perl/5\.15\.0
&nbsp;    /home/nicholas/Sandpit/snap5\.9\.x\-/lib/perl5/5\.15\.0/amd64\-freebsd\-thread\-multi
&nbsp;    /home/nicholas/Sandpit/snap5\.9\.x\-/lib/perl5/5\.15\.0
&nbsp;    \.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @iabyn

On Wed\, Jun 15\, 2011 at 04&#8203;:06&#8203;:39PM \+0100\, Nicholas Clark wrote&#8203;:
> On Wed\, Jun 15\, 2011 at 03&#8203;:59&#8203;:25PM \+0100\, Tom Hukins wrote&#8203;:
> > On Wed\, Jun 15\, 2011 at 03&#8203;:12&#8203;:29PM \+0100\, Dave Mitchell wrote&#8203;:
> > > Now fixed \(hopefully\) by the following&#8203;:
> > > 
> > > commit 6e59d93aee950461947904b4f24a7f52c6c85f58
> > 
> > This test still hangs for me in the same way as before\.  The kdump
> > output looks much like the one Nicholas posted earlier\.
> 
> Sadly\, me too\, on FreeBSD 7\.2

Ok\, I've given up for now and reverted my last two eintr\.t changes

\-\- 
"You're so sadly neglected\, and often ignored\.
A poor second to Belgium\, When going abroad\."
&nbsp;    \-\- Monty Python\, "Finland"
